opm/autodiff/CPRPreconditioner.hpp

```diff
@@ -43,8 +46,222 @@
 
 namespace Opm
 {
```

Comment:
The section below seems to contain lots of memory management-related code. I do not understand why it is necessary. Can you give a brief overview or explanation?